core/src/main/java/org/apache/iceberg/io/datafile/WriteBuilder.java

```diff
@@ -0,0 +1,340 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.io.datafile;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import org.apache.iceberg.DeleteFile;
+import org.apache.iceberg.FileFormat;
+import org.apache.iceberg.MetricsConfig;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.SortOrder;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.deletes.EqualityDeleteWriter;
+import org.apache.iceberg.deletes.PositionDeleteWriter;
+import org.apache.iceberg.encryption.EncryptionKeyMetadata;
+import org.apache.iceberg.io.DataWriter;
+import org.apache.iceberg.io.DeleteSchemaUtil;
+import org.apache.iceberg.io.FileAppender;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.util.ArrayUtil;
+
+/**
+ * Builder for generating one of the following:
+ *
+ * <ul>
+ *   <li>{@link FileAppender}
+ *   <li>{@link DataWriter}
+ *   <li>{@link EqualityDeleteWriter}
+ *   <li>{@link PositionDeleteWriter}
+ * </ul>
+ *
+ * The builder wraps the file format specific {@link AppenderBuilder}. To allow further engine and
+ * file format specific configuration changes for the given writer the {@link
+ * AppenderBuilder#build(AppenderBuilder.WriteMode)} method is called with the correct parameter to
+ * create the appender used internally by the file appender and the writers.
+ *
+ * @param <A> type of the appender
+ * @param <E> engine specific schema of the input records used for appender initialization
+ */
+public class WriteBuilder<A extends AppenderBuilder<A, E>, E> {
+  private final AppenderBuilder<A, E> appenderBuilder;
+  private final String location;
+  private final FileFormat format;
+  private PartitionSpec spec = null;
+  private StructLike partition = null;
+  private EncryptionKeyMetadata keyMetadata = null;
+  private SortOrder sortOrder = null;
+  private Schema rowSchema = null;
+  private int[] equalityFieldIds = null;
+
+  WriteBuilder(AppenderBuilder<A, E> appenderBuilder, String location, FileFormat format) {
+    this.appenderBuilder = appenderBuilder;
+    this.location = location;
+    this.format = format;
+  }
+
+  /**
+   * Sets the configurations coming from the table like {@link #schema(Schema)}, {@link #set(Map)}
+   * and {@link #metricsConfig(MetricsConfig)}.
+   */
+  public WriteBuilder<A, E> forTable(Table table) {
```

Comment:
This method was originally supposed to save some time, but because not all writers use the same schema for a table and we are still manually forwarding table settings, I don't think I would carry it forward into the new API.

Reply:
Removed/deprecated